utilities/generate_messages_catalog.py

```diff
@@ -131,96 +134,172 @@ def _is_callable(item):
+    package. This will recurse down the directory paths of any package to find
+    all modules and subpackages in order to create an exhaustive list of all
+    python files within a given package.
+
+    :param top_level_directory: Path to the top level of a python package.
+    :type top_level_directory: str
+    :return: Returns a list of paths to all python files within that package.
+    :rtype: list [str]
+    """
+    modules = []
+    directories = []
+
+    for item in os.listdir(top_level_directory):
+        if item.endswith(".py"):
+            modules.append(os.path.join(top_level_directory, item))
+        elif os.path.isdir(os.path.join(top_level_directory, item)):
+            directories.append(os.path.join(top_level_directory, item))
+
+    for directory in directories:
+        modules.extend(_find_modules_and_directories(directory))
+
+    return modules
+
+
+def generate_locations():
+    """
+    Utility function to generate the file locations for Arelle's core, pip
+    installed plugins, and non-installable plugins which have been copied to
+    the non_library_plugins folder. These locations represent a list of the
+    locations for the message generation to begin ast walks in order to find
+    and generate messages for the catalog.
+
+    :return: Returns a list of strings representing module locations
+    :rtype: list [str]
+    """
+    arelle_modules = []
+
+    arelle_src_path = os.path.dirname(arelle.__file__)
+    arelle_component_locations = [
+        arelle_src_path,
+        os.path.join(os.path.dirname(__file__), NON_LIBRARY_PLUGINS)
+    ]
+
+    arelle_component_locations.extend(_find_plugin_locations())
+
+    for location in arelle_component_locations:
+        arelle_modules.extend(_find_modules_and_directories(location))
+
+    return arelle_modules
+
+
+def _find_plugin_locations():
+    """
+    Helper utility to introspect the plugin file then find a list of all the
+    source directory of each plugin to be passed to the list of roots to use
+    during the AST.walk for message inspection
+
+    :return: Return is a list of strings representing the paths to the install
+        locations of each plugin in the plugin requirements file.
+    :rtype: list [str]
+    """
+    plugin_locations = []
+    plugin_list = list(pkutils.parse_requirements(PLUGINS_FILE))
+
+    for plugin_requirement in plugin_list:
+        # Checking for a pinned requirement in the file to remove the extra
+        # information to prevent issues with the __import__ statement.
+        pin_separation_index = plugin_requirement.find("=")
+        if pin_separation_index > 0:
+            plugin_name_only = plugin_requirement[:pin_separation_index]
+        else:
+            plugin_name_only = plugin_requirement
+        plugin = __import__(plugin_name_only, globals(), locals(), [], 0)
+        plugin_location = os.path.dirname(plugin.__file__)
+        del plugin
+        plugin_locations.append(plugin_location)
```

Comment:
should there be a break after this?  Since we have found that the file has a plugin requirement?  Or is there something else going on further up the stack that makes that superfluous?